ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/RequestDataStep.java

```diff
@@ -129,15 +130,32 @@ public CompletableFuture<List<Task<SnapDataRequest>>> requestStorage(
             (response, error) -> {
               if (response != null) {
                 downloadState.removeOutstandingTask(getStorageRangeTask);
-                for (int i = 0; i < response.slots().size(); i++) {
-                  final StorageRangeDataRequest request =
-                      (StorageRangeDataRequest) requestTasks.get(i).getData();
-                  request.setRootHash(blockHeader.getStateRoot());
-                  request.addResponse(
-                      downloadState,
-                      worldStateProofProvider,
-                      response.slots().get(i),
-                      i < response.slots().size() - 1 ? new ArrayDeque<>() : response.proofs());
+                if (response.slots().size() > 0) {
+                  for (int i = 0; i < response.slots().size(); i++) {
+                    final StorageRangeDataRequest request =
+                        (StorageRangeDataRequest) requestTasks.get(i).getData();
+                    request.setRootHash(blockHeader.getStateRoot());
+                    request.addResponse(
+                        downloadState,
+                        worldStateProofProvider,
+                        response.slots().get(i),
+                        i < response.slots().size() - 1 ? new ArrayDeque<>() : response.proofs());
+                  }
+                }
+
+                // if we received no slots, but do do get a proof of exclusion, add as a response:
+                else if (response.proofs().size() > 0 && requestTasks.size() > 0) {
+                  requestTasks.stream()
+                      .findFirst()
```

Comment:
not clear for me why we are doing that only for the first one . it it because the server will stop at the first missing request ?

Reply:
added response for all